### PR TITLE
Refactor and add tests coverage on package runconfig

### PR DESCRIPTION
**Pull request by vdemeester**
After [opts](https://github.com/docker/docker/pull/13694), this is another big one — probably a little too big.. It's the time for `runconfig` to get covered 🐸.

What this PR does is :
- A small re-organization of existing tests (putting the parsing one into `parse_test.go`, the merging one into `merge_test.go`, and so on).
- Add `compare_test.go` for more extensive testing of the `Compare` config method.
- Update `config_test.go` to test `Entrypoint` and `Command` structs (that might be overkill but… oh well :smiley_cat:).
- Add `exec_test.go` to test `ParseExec` (which is in `exec.go`).
- Add `hostconfig_test.go` for `hostconfig.go` tests, with fixtures (similar to what was done with `config_test.go`).
- Update `parse_test.go` to cover more cases. As some of some parse thingy are calling the `opts` package and some aren't I tried to limit test case for those shared validation cases. I'll use the _data gathered_ by those PR to update the issue I created about the need to centralize some of the checks (https://github.com/docker/docker/issues/13824).

There is a few `FIXME` in there, question I was asking myself and not really able to answer. I'll comment on them directly in this PR.

``` bash
# Before
+ go test -test.timeout=30m github.com/docker/docker/runconfig
PASS
coverage: 54.6% of statements
# After
+ go test -test.timeout=30m github.com/docker/docker/runconfig
PASS
coverage: 94.8% of statements
```

It's probably gonna be a little tricky to merge, because some PRs might changes these files a bit (like https://github.com/docker/docker/pull/14133) — but I'll rebase, fix or remove tests if needed (or even split this PR into severals).

Thanks for the reviews :)

Signed-off-by: Vincent Demeester vincent@sbr.pm

### CODE REVIEW

File: runconfig/parse_test.go [hunk 1]
```diff
@@ -56,10 +227,336 @@ func TestNetHostname(t *testing.T) {
...
+func TestParseRestartPolicy(t *testing.T) {
+	invalids := map[string]string{
+		"something":          "invalid restart policy something",
+		"always:2":           "maximum restart count not valid with restart policy of \"always\"",
+		"on-failure:invalid": `strconv.ParseInt: parsing "invalid": invalid syntax`,
+	}
+	valids := map[string]RestartPolicy{
+		"": RestartPolicy{},
+		// FIXME This feels not right
+		"always:1:2": RestartPolicy{
```
Comment (vdemeester) [inline runconfig/parse_test.go]: I never tried with the ``cli`` but it's possible to specify this for the restart policies. I will passe the parsing and the ``MaximumRetryCount`` will be ``0``. Felt a little weird to me :sweat:.

Comment (vdemeester) [inline runconfig/parse_test.go]: It felt weird that ``always:2`` was failing (expected) but not ``always:2:1``.

File: runconfig/parse_test.go [hunk 2]
```diff
@@ -56,10 +227,335 @@ func TestNetHostname(t *testing.T) {
...
+			Name:              "always",
+			MaximumRetryCount: 0,
+		},
+		"on-failure:1": RestartPolicy{
+			Name:              "on-failure",
+			MaximumRetryCount: 1,
+		},
+		// FIXME This doesn't feel right
+		"on-failure:1:2": RestartPolicy{
```
Comment (vdemeester) [inline runconfig/parse_test.go]: This one is a little bit different but felt weird too. I thought first it wouldn't be valid but it is — and more strange, the ``MaximumRetryCount`` ends up beeing ``0``.

File: runconfig/parse_test.go [hunk 3]
```diff
@@ -56,10 +227,335 @@ func TestNetHostname(t *testing.T) {
...
+func TestParseRestartPolicy(t *testing.T) {
+	invalids := map[string]string{
+		"something":          "invalid restart policy something",
+		"always:2":           "maximum restart count not valid with restart policy of \"always\"",
+		"on-failure:invalid": `strconv.ParseInt: parsing "invalid": invalid syntax`,
+	}
+	valids := map[string]RestartPolicy{
+		"": {},
+		// FIXME This feels not right
+		"always:1:2": {
```
Comment (vdemeester) [inline runconfig/parse_test.go]: It felt weird that always:2 was failing (expected) but not always:2:1.

File: runconfig/parse_test.go [hunk 4]
```diff
@@ -56,10 +227,335 @@ func TestNetHostname(t *testing.T) {
...
+			Name:              "always",
+			MaximumRetryCount: 0,
+		},
+		"on-failure:1": {
+			Name:              "on-failure",
+			MaximumRetryCount: 1,
+		},
+		// FIXME This doesn't feel right
```
Comment (vdemeester) [inline runconfig/parse_test.go]: This one is a little bit different but felt weird too. I thought first it wouldn't be valid but it is — and more strange, the ``MaximumRetryCount`` ends up beeing ``0``.

**Comment by calavera** [comment]
LGTM

**Comment by vdemeester** [comment]
Just rebased it to fix conflicts with the move of package ``nat`` to ``pkg/nat``.

File: runconfig/parse_test.go (hunk 4 above)
Comment (LK4D4) [inline runconfig/parse_test.go]: Yeah, that's weird, pls create issue about this.

**Comment by LK4D4** [comment]
@vdemeester Pls create issues about weird things that you've found.
LGTM

**Comment by vdemeester** [comment]
Will do :wink:

**Pull request merged**